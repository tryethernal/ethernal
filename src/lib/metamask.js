const ethers = require('ethers');

export const sendTransaction = ({ ethereum, address, abi, signature, params, options }) => {
    return new Promise((resolve, reject) => {
        const provider = new ethers.providers.Web3Provider(ethereum, 'any');
        const signer = provider.getSigner();
        const contract = new ethers.Contract(address, abi, signer);
        contract.populateTransaction[signature](...params, options)
            .then(transaction => {
                const params = {
                    ...transaction,
                    value: transaction.value.toHexString()
                };
                window.ethereum.request({
                    method: 'eth_sendTransaction',
                    params: [params]
                })
                .then(resolve)
                .catch(reject);
            })
            .catch(reject);
    });
};
