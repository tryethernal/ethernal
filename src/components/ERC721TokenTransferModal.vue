<template>
<v-dialog v-model="dialog" max-width="600">
    <v-card>
        <v-card-title class="headline">Transfer {{ token.attributes.name }}</v-card-title>
        <v-form v-model="validForm">
            <v-card-text>
                <v-alert type="success" v-if="successMessage" v-html="successMessage"></v-alert>
                <v-alert type="error" v-if="errorMessage" v-html="errorMessage"></v-alert>
                <v-alert type="error" v-if="invalidOwner">The connected account is not the owner of this token.</v-alert>
                <v-alert v-if="!successMessage && !errorMessage && !isPublicExplorer && !invalidOwner" type="info" text v-html="'This will only work if your node supports either <code>hardhat_impersonateAccount</code> or <code>evm_unlockUnknownAccount</code>.'"></v-alert>

                Owner: <Hash-Link :type="'address'" :fullHash="true" :hash="token.owner"></Hash-Link>
                <v-text-field
                    v-model="recipient"
                    :rules="[v => !!v && v.length == 42 || 'Invalid address (must be 42 characters long)']"
                    small
                    outlined
                    dense
                    prepend-inner-icon="mdi-arrow-right"
                    class="mt-3"
                    id="recipient"
                    label="Recipient Address">
                </v-text-field>
                <span class="align-right" v-if="transaction.hash">
                    Transaction:
                    <v-progress-circular v-if="transaction.hash && transaction.receipt.status === undefined" class="mr-2" size="16" width="2" indeterminate color="primary"></v-progress-circular>
                    <template v-else>
                        <v-icon small v-show="transaction.receipt.status" color="success lighten-1" class="mr-1 align-with-text">mdi-check-circle</v-icon>
                        <v-icon small v-show="!transaction.receipt.status" color="error lighten-1" class="mr-1 align-with-text">mdi-alert-circle</v-icon>
                    </template>
                    <Hash-Link :type="'transaction'" :hash="transaction.hash"></Hash-Link>
                </span>
                <Metamask v-if="dialog && isPublicExplorer" class="mt-1" @rpcConnectionStatusChanged="onRpcConnectionStatusChanged"></Metamask>
            </v-card-text>

            <v-card-actions>
                <v-spacer></v-spacer>
                <v-btn color="primary" text @click.stop="close()">Close</v-btn>
                <v-btn id="transferToken" color="primary" :disabled="!validForm || isPublicExplorer && !metamaskData.isReady || invalidOwner" :loading="loading" text @click.stop="transferToken()">Transfer Token</v-btn>
            </v-card-actions>
        </v-form>
    </v-card>
</v-dialog>
</template>
<script>
import { mapGetters } from 'vuex';
import ERC721_ABI from '../abis/erc721.json';
import { sendTransaction } from '../lib/metamask';
import HashLink from './HashLink';
import Metamask from './Metamask';

export default {
    name: 'ERC721TokenTransferModal',
    props: ['address', 'token'],
    components: {
        HashLink,
        Metamask
    },
    data: () => ({
        validForm: false,
        recipient: null,
        dialog: false,
        resolve: null,
        reject: null,
        successMessage: null,
        errorMessage: null,
        loading: false,
        options: {},
        transaction: {
            receipt: {}
        },
        rpcConnectionStatus: false,
        metamaskData: {},
        invalidOwner: false,
        didTransfer: false
    }),
    methods: {
        transferWithMetamask() {
            sendTransaction({
                ethereum: window.ethereum,
                address: this.address,
                abi: ERC721_ABI,
                signature: 'safeTransferFrom(address,address,uint256)',
                params: [this.metamaskData.account, this.recipient, this.token.tokenId],
                options: { value: '0' }
            })
            .then(txHash => {
                this.successMessage = `Transfer transaction sent: <a class="white--text" href="/transaction/${txHash}">${txHash}</a>`;
                this.didTransfer = true;
            })
            .catch(error => this.errorMessage = error.message || error)
            .finally(() => this.loading = false);
        },
        onRpcConnectionStatusChanged(data) {
            this.metamaskData = data;
            this.invalidOwner = data.isReady && this.token.owner !== data.account;
        },
        transferToken() {
            this.loading = true;
            this.successMessage = null;
            this.errorMessage = null;
            if (this.metamaskData.isReady)
                return this.transferWithMetamask();

            this.server.impersonateAccount(this.currentWorkspace.rpcServer, this.token.owner)
                .then(hasBeenUnlocked => {
                    if (!hasBeenUnlocked)
                        throw new Error("Transfer failed. Couldn't unlock owner account.");

                    this.server.transferErc721Token(this.currentWorkspace.rpcServer, this.address, this.token.owner, this.recipient, this.token.tokenId)
                        .then(transaction => {
                            this.transaction = { ...transaction, receipt: {} };
                            transaction.wait().then(receipt => {
                                this.transaction.receipt = receipt;
                                if (receipt.status) {
                                    this.successMessage = `Token transferred successfully!`;
                                    this.didTransfer = true;
                                }
                                else {
                                    console.log(transaction, receipt);
                                    this.errorMessage = `Transaction failed`;
                                }
                            });
                        })
                        .catch(error => {
                            console.log(error);
                            this.errorMessage = error.reason || `Error: ${error.message}` || 'Error while sending the transaction. Please reload the page and try again.';
                        })
                        .finally(() => this.loading = false);
                })
                .catch(error => {
                    console.log(error);
                    this.errorMessage = error.message || error;
                    this.loading = false;
                })
        },
        open(options) {
            this.dialog = true;
            this.options = options || {};
            return new Promise((resolve, reject) => {
                this.resolve = resolve;
                this.reject = reject;
            });
        },
        close() {
            this.resolve(this.didTransfer);
            this.reset();
        },
        reset() {
            this.options = {};
            this.transaction = {
                receipt: {}
            };
            this.validForm = false;
            this.didTransfer = false;
            this.rpcConnectionStatus = false;
            this.metamaskData = {};
            this.invalidOwner = false;
            this.recipient = null;
            this.successMessage = null;
            this.errorMessage = null;
            this.loading = false;
            this.dialog = false;
            this.resolve = null;
            this.reject = null;
        }
    },
    computed: {
        ...mapGetters([
            'currentWorkspace',
            'isPublicExplorer'
        ])
    }
}
</script>
