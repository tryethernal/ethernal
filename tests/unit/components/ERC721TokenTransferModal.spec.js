import MockHelper from '../MockHelper';
import ethereum from '../mocks/ethereum';
import '../mocks/metamask';

window.ethereum = ethereum;
  
import ERC721TokenTransferModal from '@/components/ERC721TokenTransferModal.vue';

describe('ERC721TokenTransferModal.vue', () => {
    let helper;

    beforeEach(() => {
        helper = new MockHelper();
    });

    it('Should not show the metamask button if not public explorer', async () => {
        const wrapper = helper.mountFn(ERC721TokenTransferModal, {
            propsData: {
                address: '0x123',
                token: {
                    owner: '0xabc',
                    attributes: { name: 'My Token #1' }
                }
            },
            stubs: ['Hash-Link', 'Metamask']
        });
        await wrapper.setData({ dialog: true, options: {} });

        await new Promise(process.nextTick);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should successfully impersonate the account & call the transfer function', async () => {
        jest.spyOn(helper.mocks.server, 'impersonateAccount').mockResolvedValue(true);
        jest.spyOn(helper.mocks.server, 'transferErc721Token').mockResolvedValue({
            hash: '0xdef',
            wait: jest.fn().mockResolvedValue({ status: 1 })
        });
        const wrapper = helper.mountFn(ERC721TokenTransferModal, {
            propsData: {
                address: '0x123',
                token: {
                    owner: '0xabc',
                    attributes: { name: 'My Token #1' }
                }
            },
            stubs: ['Hash-Link', 'Metamask']
        });
        await wrapper.setData({ dialog: true, options: {} });
        await wrapper.find('#recipient').setValue('0x29ea412cc10a9cfc08c2298f382b2fe01e6ca83b');
        await new Promise(process.nextTick);

        await wrapper.find('#transferToken').trigger('click');
        await new Promise(process.nextTick);

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should show the Metamask button if public explorer', async () => {
        const wrapper = helper.mountFn(ERC721TokenTransferModal, {
            propsData: {
                address: '0x123',
                token: {
                    owner: '0xabc',
                    attributes: { name: 'My Token #1' }
                }
            },
            stubs: ['Hash-Link', 'Metamask'],
            getters: {
                isPublicExplorer: jest.fn().mockReturnValue(true)
            }
        });

        await wrapper.setData({ dialog: true, options: {}, metamaskData: { account: '0x456' }, invalidOwner: false });
        await new Promise(process.nextTick);

        expect(wrapper.html()).toMatchSnapshot();
    });


    it('Should send a transation with Metamask', async () => {
        const wrapper = helper.mountFn(ERC721TokenTransferModal, {
            propsData: {
                address: '0x123',
                token: {
                    owner: '0xabc',
                    attributes: { name: 'My Token #1' }
                }
            },
            stubs: ['Hash-Link', 'Metamask'],
            getters: {
                isPublicExplorer: jest.fn().mockReturnValue(true)
            }
        });

        await wrapper.setData({ dialog: true, options: {}, metamaskData: { account: '0x456', isReady: true }, invalidOwner: false });
        await wrapper.find('#recipient').setValue('0x29ea412cc10a9cfc08c2298f382b2fe01e6ca83b');
        await new Promise(process.nextTick);

        await wrapper.find('#transferToken').trigger('click');
        await new Promise(process.nextTick);

        expect(wrapper.html()).toMatchSnapshot();
    });
});
