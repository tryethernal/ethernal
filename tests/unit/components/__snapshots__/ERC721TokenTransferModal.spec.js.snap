// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ERC721TokenTransferModal.vue Should not show the metamask button if not public explorer 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 600px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Transfer My Token #1</div>
          <form novalidate="novalidate" class="v-form">
            <div class="v-card__text">
              <!---->
              <!---->
              <!---->
              <div role="alert" class="v-alert v-sheet theme--dark v-alert--text info--text">This will only work if your node supports either <code>hardhat_impersonateAccount</code> or <code>evm_unlockUnknownAccount</code>.</div>

              Owner: <hash-link-stub fullhash="true" hash="0xabc" type="address"></hash-link-stub>
              <div class="v-input mt-3 v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-input__prepend-inner">
                      <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate mdi mdi-arrow-right theme--light"></i></div>
                    </div>
                    <fieldset aria-hidden="true">
                      <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="recipient" class="v-label theme--light" style="left: 0px; position: absolute;">Recipient Address</label><input small="" id="recipient" type="text"></div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                    </div>
                  </div>
                </div>
              </div>
              <!---->
              <!---->
            </div>
            <div class="v-card__actions">
              <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--text theme--light v-size--default" id="transferToken"><span class="v-btn__content">Transfer Token</span></button>
            </div>
          </form>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`ERC721TokenTransferModal.vue Should send a transation with Metamask 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 600px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Transfer My Token #1</div>
          <form novalidate="novalidate" class="v-form">
            <div class="v-card__text">
              <div role="alert" class="v-alert v-sheet theme--dark success">Transfer transaction sent: <a class="white--text" href="/transaction/0x1234">0x1234</a></div>
              <!---->
              <!---->
              <!---->

              Owner: <hash-link-stub fullhash="true" hash="0xabc" type="address"></hash-link-stub>
              <div class="v-input mt-3 v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-input__prepend-inner">
                      <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate mdi mdi-arrow-right theme--light"></i></div>
                    </div>
                    <fieldset aria-hidden="true">
                      <legend style="width: -24px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="recipient" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Recipient Address</label><input small="" id="recipient" type="text"></div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                    </div>
                  </div>
                </div>
              </div>
              <!---->
              <metamask-stub class="mt-1"></metamask-stub>
            </div>
            <div class="v-card__actions">
              <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text" id="transferToken"><span class="v-btn__content">Transfer Token</span></button>
            </div>
          </form>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`ERC721TokenTransferModal.vue Should show the Metamask button if public explorer 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 600px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Transfer My Token #1</div>
          <form novalidate="novalidate" class="v-form">
            <div class="v-card__text">
              <!---->
              <!---->
              <!---->
              <!---->

              Owner: <hash-link-stub fullhash="true" hash="0xabc" type="address"></hash-link-stub>
              <div class="v-input mt-3 v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-input__prepend-inner">
                      <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate mdi mdi-arrow-right theme--light"></i></div>
                    </div>
                    <fieldset aria-hidden="true">
                      <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="recipient" class="v-label theme--light" style="left: 0px; position: absolute;">Recipient Address</label><input small="" id="recipient" type="text"></div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                    </div>
                  </div>
                </div>
              </div>
              <!---->
              <metamask-stub class="mt-1"></metamask-stub>
            </div>
            <div class="v-card__actions">
              <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--text theme--light v-size--default" id="transferToken"><span class="v-btn__content">Transfer Token</span></button>
            </div>
          </form>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`ERC721TokenTransferModal.vue Should successfully impersonate the account & call the transfer function 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 600px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Transfer My Token #1</div>
          <form novalidate="novalidate" class="v-form">
            <div class="v-card__text">
              <div role="alert" class="v-alert v-sheet theme--dark success">Token transferred successfully!</div>
              <!---->
              <!---->
              <!---->

              Owner: <hash-link-stub fullhash="true" hash="0xabc" type="address"></hash-link-stub>
              <div class="v-input mt-3 v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-input__prepend-inner">
                      <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate mdi mdi-arrow-right theme--light"></i></div>
                    </div>
                    <fieldset aria-hidden="true">
                      <legend style="width: -24px;"><span class="notranslate">​</span></legend>
                    </fieldset>
                    <div class="v-text-field__slot"><label for="recipient" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Recipient Address</label><input small="" id="recipient" type="text"></div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                    </div>
                  </div>
                </div>
              </div> <span class="align-right">
                    Transaction:
                    <i aria-hidden="true" class="v-icon notranslate mr-1 align-with-text mdi mdi-check-circle theme--light success--text text--lighten-1" style="font-size: 16px;"></i> <i aria-hidden="true" class="v-icon notranslate mr-1 align-with-text mdi mdi-alert-circle theme--light error--text text--lighten-1" style="font-size: 16px; display: none;"></i> <hash-link-stub hash="0xdef" type="transaction"></hash-link-stub></span>
              <!---->
            </div>
            <div class="v-card__actions">
              <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text" id="transferToken"><span class="v-btn__content">Transfer Token</span></button>
            </div>
          </form>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;
